src/docfx/build/toc/ServicePageGenerator.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// Generate Service Page.
+    /// </summary>
+    internal class ServicePageGenerator
+    {
+        private readonly Input _input;
+        private readonly JoinTOCConfig _joinTOCConfig;
+        private readonly RepositoryProvider _repositoryProvider;
+        private readonly Output _output;
+
+        public ServicePageGenerator(
+            Input input,
+            JoinTOCConfig joinTOCConfig,
+            RepositoryProvider repositoryProvider,
+            Output output)
+        {
+            _input = input;
+            _joinTOCConfig = joinTOCConfig;
+            _repositoryProvider = repositoryProvider;
+            _output = output;
+        }
+
+        public void GenerateServicePageFromTopLevelTOC(
+            TableOfContentsNode node,
+            ConcurrentBag<FilePath> results,
+            string directoryName = "")
+        {
+            if (node == null || string.IsNullOrEmpty(node.Name))
+            {
+                return;
+            }
+
+            var filename = Regex.Replace(node.Name, @"\s+", "");
+
+            if (!string.IsNullOrEmpty(node.LandingPageType.ToString()))
+            {
+                var baseDir = string.IsNullOrEmpty(_joinTOCConfig.OutputPath) ? Directory.GetCurrentDirectory() : _joinTOCConfig.OutputPath.Replace("..", "");
+                var servicePagePath = FilePath.Generated(new PathString($"./{baseDir}/{directoryName}/{filename}.yml"));
+
+                var name = node.Name;
+                var fullName = node.Name;
+                var uid = node.Uid;
+                var href = node.Href;
+
+                var children = new List<SourceInfo<ServicePageModel>>();
+                foreach (var item in node.Items)
+                {
+                    ServicePageModel newItem;
+                    var newItemName = item.Value.Name;
+                    var newItemHref = item.Value.Href;
+                    var newItemUid = item.Value.Uid;
+
+                    if (!string.IsNullOrEmpty(newItemHref.Value))
+                    {
+                        if (_repositoryProvider.Repository != null && item.Source != null)
```

Comment:
The href resolve should work regardless of repository and the current algorithm isn't quite right.

Reply:
Since we need to get the full path of output folder, which is related to repository root path. Once `.openpublishing.publish.config.json` exists, there should be a repository.

Reply:
Now the href depends `docsetPath`， should work now.

Reply:
Finished.